sound/soc/sof/pcm.c

```diff
@@ -336,21 +333,33 @@ static int sof_pcm_trigger(struct snd_soc_component *component,
 		return -EINVAL;
 	}
 
-	/*
-	 * DMA and IPC sequence is different for start and stop. Need to send
-	 * STOP IPC before stop DMA
-	 */
-	if (!ipc_first)
-		snd_sof_pcm_platform_trigger(sdev, substream, cmd);
+	/* DAI DMA pre-trigger actions */
+	if (pcm_ops && pcm_ops->pre_trigger)
+		ret = pcm_ops->pre_trigger(component, substream, cmd);
 
-	if (pcm_ops && pcm_ops->trigger)
-		ret = pcm_ops->trigger(component, substream, cmd);
-
-	/* need to STOP DMA even if trigger IPC failed */
-	if (ipc_first)
+	switch (cmd) {
+	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+	case SNDRV_PCM_TRIGGER_START:
+		/* invoke platform trigger to start DMA only if pre_trigger is successful */
+		if (ret < 0)
+			return ret;
 		snd_sof_pcm_platform_trigger(sdev, substream, cmd);
+		break;
+	case SNDRV_PCM_TRIGGER_SUSPEND:
+	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+	case SNDRV_PCM_TRIGGER_STOP:
+		/* invoke platform trigger to stop DMA even if pre_trigger failed */
```

Comment:
I am just too confused here:

The comment on like 336 kinda hints that the DAI controls the DMA, but here it's the platform.

The commit message is equally convoluted. Everything seems to have triggers.

Lost....

Reply:
@plbossart I should have done this a long time ago when we did the split for the HDA DAI ops. What this PR does is just mirror the same thing for the host DMA part.

Reply:
it's even less clear to me what the comments referring to 'dai' mean if this is related to the HOST DMA management.

Reply:
The DAI refers to the FE CPU DAI. I can change it to the Host DMA instead of DAI DMA in the comments if it makes it clearer.